pallets/evm-fungible-app/src/lib.rs

```diff
@@ -0,0 +1,839 @@
+                    target,
+                    max_gas: 2000000u64.into(),
+                },
+            )?;
+            Ok(())
+        }
+
+        #[pallet::call_index(6)]
+        #[pallet::weight(<T as Config>::WeightInfo::register_fungible_app())]
+        pub fn register_fungible_app(
+            origin: OriginFor<T>,
+            network_id: EVMChainId,
+            contract: H160,
+        ) -> DispatchResult {
+            ensure_root(origin)?;
+            ensure!(
+                !AppAddresses::<T>::contains_key(network_id, AssetKind::Thischain),
+                Error::<T>::AppAlreadyRegistered
+            );
+            ensure!(
+                !AppAddresses::<T>::contains_key(network_id, AssetKind::Sidechain),
+                Error::<T>::AppAlreadyRegistered
+            );
+            AppAddresses::<T>::insert(network_id, AssetKind::Thischain, contract);
+            AppAddresses::<T>::insert(network_id, AssetKind::Sidechain, contract);
+            T::AppRegistry::register_app(network_id, contract)?;
+            Ok(())
+        }
+
+        #[pallet::call_index(7)]
+        #[pallet::weight(<T as Config>::WeightInfo::register_native_app())]
+        pub fn register_native_app(
+            origin: OriginFor<T>,
+            network_id: EVMChainId,
+            contract: H160,
+            asset_id: AssetIdOf<T>,
+            sidechain_precision: u8,
+        ) -> DispatchResult {
+            ensure_root(origin)?;
+            ensure!(
+                !AppAddresses::<T>::contains_key(network_id, AssetKind::Native),
+                Error::<T>::AppAlreadyRegistered
+            );
+            AppAddresses::<T>::insert(network_id, AssetKind::Native, contract);
+            NativeAssetIds::<T>::insert(network_id, &asset_id);
+            SidechainPrecision::<T>::insert(network_id, &asset_id, sidechain_precision);
+            T::AppRegistry::register_app(network_id, contract)?;
+            Ok(())
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        pub fn register_asset_inner(
+            network_id: EVMChainId,
+            asset_id: AssetIdOf<T>,
+            contract: H160,
+            asset_kind: AssetKind,
+            sidechain_precision: u8,
+        ) -> DispatchResult {
+            ensure!(
+                AppAddresses::<T>::contains_key(network_id, asset_kind),
+                Error::<T>::AppIsNotRegistered
+            );
+            ensure!(
+                !TokenAddresses::<T>::contains_key(network_id, &asset_id),
+                Error::<T>::TokenAlreadyRegistered
+            );
+            TokenAddresses::<T>::insert(network_id, &asset_id, contract);
+            AssetsByAddresses::<T>::insert(network_id, contract, &asset_id);
+            AssetKinds::<T>::insert(network_id, &asset_id, asset_kind);
+            SidechainPrecision::<T>::insert(network_id, &asset_id, sidechain_precision);
+            T::AssetRegistry::manage_asset(network_id.into(), asset_id)?;
```

Comment:
May we need a special event for registering asset?

Reply:
Added